File: src/apps/web/src/data/chartDrawings.ts
/** Pure chart-drawing storage and synchronization primitives. */

export type DrawingTime = string | number | { year: number; month: number; day: number }
export interface ChartDrawingPoint { time: DrawingTime; price: number }
export interface ChartDrawing { id: string; tool: string; points: ChartDrawingPoint[]; origin_timeframe?: string; cross_timeframe?: boolean; revision?: number }
export type DrawingPersistence = 'syncing' | 'synced' | 'device-only' | 'conflict' | 'failed'
export interface DrawingScope { userId: number; market: 'US' | 'CN'; symbol: string; timeframe: string; crossTimeframe: boolean }
export interface CachedDrawings { drawings: ChartDrawing[]; savedAt: number }
export interface DrawingTombstone { drawing: ChartDrawing; revision: number; origin_timeframe: string; cross_timeframe: boolean }
export interface DrawingTombstoneRevision { drawing_id: string; revision: number; origin_timeframe: string; cross_timeframe: boolean }
export interface ConflictRefreshDecisionInput {
  capturedScopeToken: string
  currentScopeToken: string
  capturedRequestGeneration: number
  currentRequestGeneration: number
  currentEditGeneration: number
  refreshed: ChartDrawing[]
  latest: ChartDrawing[]
  fallback: Pick<DrawingScope, 'timeframe' | 'crossTimeframe'>
  tombstones?: DrawingTombstone[]
  refreshedTombstones?: DrawingTombstoneRevision[]
  previousRetryKey: string | null
}
export type DrawingOperation =
  | { op: 'upsert'; origin_timeframe: string; cross_timeframe: boolean; revision: number | null; drawing: ChartDrawing }
  | { op: 'delete' | 'restore'; origin_timeframe: string; cross_timeframe: boolean; revision: number; drawing_id: string }

const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
const TIMEFRAME = /^[\p{L}\p{N}_-]{1,16}$/u
const DATE = /^\d{4}-\d{2}-\d{2}$/
const TOOL_POINTS: Record<string, number> = {
  segment: 2, horizontal: 1, 'horizontal-segment': 2, vertical: 1, ray: 2, straight: 2, parallel: 3, channel: 3, periodic: 2, 'info-line': 2, 'smooth-top': 3, cross: 1,
  rectangle: 2, triangle: 3, parallelogram: 3, circle: 2, ellipse: 2, path: 5,
  wave3: 4, wave5: 6, wave8: 9, 'head-shoulders': 5, 'triangle-pattern': 5, mw: 5, abcd: 4, xabcd: 5, 'three-drive': 7, sine: 6,
  'fib-retracement': 2, 'fib-time': 2, 'fib-extension': 3, 'speed-resistance': 2, 'gann-box': 2, 'gann-angle': 2, 'grid-line': 2, pitchfork: 3, schiff: 3, 'modified-schiff': 3, 'inside-pitchfork': 3, fan: 2,
  'time-ruler': 2, 'space-ruler': 2, 'time-space-ruler': 2, 'long-position': 2, 'short-position': 2, 'price-label': 1, arrow: 2, 'up-arrow': 2, 'down-arrow': 2,
}

function isCalendarDate(value: string) {
  if (!DATE.test(value)) return false
  const parsed = new Date(`${value}T00:00:00Z`)
  return !Number.isNaN(parsed.valueOf()) && parsed.toISOString().slice(0, 10) === value
}

export function chartDrawingKey(scope: DrawingScope) {
  return `ciclotrade:drawings:v3:${scope.userId}:${scope.market}:${scope.symbol.toUpperCase()}:${scope.timeframe}:${scope.crossTimeframe ? 'all' : 'current'}`
}

export function legacyChartDrawingKey(symbol: string, timeframe: string, crossTimeframe: boolean) {
  return `ciclotrade:drawings:v2:${symbol.toUpperCase()}:${crossTimeframe ? 'all' : timeframe}`
}

export function hasLegacyChartDrawings(storage: Pick<Storage, 'getItem'>, scope: Pick<DrawingScope, 'symbol' | 'timeframe' | 'crossTimeframe'>) {
  return storage.getItem(legacyChartDrawingKey(scope.symbol, scope.timeframe, scope.crossTimeframe)) !== null
}

export function drawingScopeToken(scope: Pick<DrawingScope, 'userId' | 'market' | 'symbol' | 'timeframe' | 'crossTimeframe'>) {
  return chartDrawingKey(scope)
}

export function isCurrentDrawingScope(token: string, scope: DrawingScope) {
  return token === drawingScopeToken(scope)
}

export function isValidChartDrawing(value: unknown): value is ChartDrawing {
  if (!value || typeof value !== 'object') return false
  const drawing = value as Partial<ChartDrawing>
  if (Object.keys(drawing).some((key) => !['id', 'tool', 'points', 'origin_timeframe', 'cross_timeframe', 'revision'].includes(key))) return false
  if (typeof drawing.id !== 'string' || !UUID.test(drawing.id) || typeof drawing.tool !== 'string' || TOOL_POINTS[drawing.tool] === undefined || !Array.isArray(drawing.points) || drawing.points.length !== TOOL_POINTS[drawing.tool]) return false
  if (drawing.origin_timeframe !== undefined && (typeof drawing.origin_timeframe !== 'string' || !TIMEFRAME.test(drawing.origin_timeframe))) return false
  if (drawing.cross_timeframe !== undefined && typeof drawing.cross_timeframe !== 'boolean') return false
  if (drawing.revision !== undefined && (!Number.isInteger(drawing.revision) || drawing.revision < 1)) return false
  return drawing.points.every((point) => {
    if (!point || typeof point !== 'object' || Object.keys(point).some((key) => key !== 'time' && key !== 'price')) return false
    const typed = point as Partial<ChartDrawingPoint>
    const businessDay = typed.time && typeof typed.time === 'object' && Object.keys(typed.time).every((key) => ['year', 'month', 'day'].includes(key))
      && Number.isInteger(typed.time.year) && Number.isInteger(typed.time.month) && Number.isInteger(typed.time.day)
      && typed.time.year >= 1970 && typed.time.year <= 2100 && typed.time.month >= 1 && typed.time.month <= 12 && typed.time.day >= 1 && typed.time.day <= 31
      && new Date(Date.UTC(typed.time.year, typed.time.month - 1, typed.time.day)).getUTCFullYear() === typed.time.year
      && new Date(Date.UTC(typed.time.year, typed.time.month - 1, typed.time.day)).getUTCMonth() === typed.time.month - 1
      && new Date(Date.UTC(typed.time.year, typed.time.month - 1, typed.time.day)).getUTCDate() === typed.time.day
    return ((typeof typed.time === 'number' && Number.isSafeInteger(typed.time) && typed.time > 0 && typed.time < 4_102_444_800) || (typeof typed.time === 'string' && isCalendarDate(typed.time)) || businessDay)
      && typeof typed.price === 'number' && Number.isFinite(typed.price) && Math.abs(typed.price) <= 1_000_000_000
  })
}

export function readCachedDrawings(storage: Pick<Storage, 'getItem'>, scope: DrawingScope): CachedDrawings | null {
  try {
    const value = JSON.parse(storage.getItem(chartDrawingKey(scope)) ?? 'null') as CachedDrawings | null
    return value && Array.isArray(value.drawings) && value.drawings.length <= 200 && value.drawings.every(isValidChartDrawing) && Number.isFinite(value.savedAt) ? value : null
  } catch { return null }
}

export function writeCachedDrawings(storage: Pick<Storage, 'setItem'>, scope: DrawingScope, drawings: ChartDrawing[]) {
  if (drawings.length > 200 || !drawings.every(isValidChartDrawing)) throw new Error('本机画线数据无效。')
  storage.setItem(chartDrawingKey(scope), JSON.stringify({ drawings, savedAt: Date.now() } satisfies CachedDrawings))
}

function comparable(drawing: ChartDrawing) {
  return JSON.stringify({ id: drawing.id, tool: drawing.tool, points: drawing.points })
}

function wireDrawing(drawing: ChartDrawing): ChartDrawing {
  return { id: drawing.id, tool: drawing.tool, points: drawing.points }
}

function drawingScope(drawing: ChartDrawing, fallback: Pick<DrawingScope, 'timeframe' | 'crossTimeframe'>) {
  return { origin_timeframe: drawing.origin_timeframe ?? fallback.timeframe, cross_timeframe: drawing.cross_timeframe ?? fallback.crossTimeframe }
}

type DrawingOperationIdentity =
  | { drawing_id: string; origin_timeframe: string; cross_timeframe: boolean }
  | { drawing: Pick<ChartDrawing, 'id'>; origin_timeframe: string; cross_timeframe: boolean }

export function drawingOperationKey(value: DrawingOperationIdentity) {
  const drawingId = 'drawing_id' in value ? value.drawing_id : value.drawing.id
  return `${value.origin_timeframe}:${value.cross_timeframe ? '1' : '0'}:${drawingId}`
}

export function drawingOperations(previous: ChartDrawing[], next: ChartDrawing[], fallback: Pick<DrawingScope, 'timeframe' | 'crossTimeframe'>, tombstones: DrawingTombstone[] = []): DrawingOperation[] {
  if (next.length > 200 || !previous.every(isValidChartDrawing) || !next.every(isValidChartDrawing)) throw new Error('画线数据无效或超过上限。')
  const scopedId = (drawing: ChartDrawing) => {
    const location = drawingScope(drawing, fallback)
    return `${location.origin_timeframe}:${location.cross_timeframe ? '1' : '0'}:${drawing.id}`
  }
  const oldById = new Map(previous.map((drawing) => [scopedId(drawing), drawing]))
  const nextById = new Map(next.map((drawing) => [scopedId(drawing), drawing]))
  const deletedById = new Map(tombstones.map((tombstone) => [
    `${tombstone.origin_timeframe}:${tombstone.cross_timeframe ? '1' : '0'}:${tombstone.drawing.id}`,
    tombstone,
  ]))
  const operations: DrawingOperation[] = []
  for (const drawing of next) {
    const before = oldById.get(scopedId(drawing))
    const location = drawingScope(drawing, fallback)
    const tombstone = deletedById.get(scopedId(drawing))
    if (!before && tombstone) operations.push({ op: 'restore', ...location, revision: tombstone.revision, drawing_id: drawing.id })
    else if (!before) operations.push({ op: 'upsert', ...location, revision: null, drawing: wireDrawing(drawing) })
    else if (comparable(before) !== comparable(drawing)) operations.push({ op: 'upsert', ...location, revision: before.revision ?? null, drawing: wireDrawing(drawing) })
  }
  for (const drawing of previous) {
    if (!nextById.has(scopedId(drawing))) {
      if (!drawing.revision) throw new Error('本机离线画线尚未同步，不能删除。')
      operations.push({ op: 'delete', ...drawingScope(drawing, fallback), revision: drawing.revision, drawing_id: drawing.id })
    }
  }
  return operations
}

function remoteRevisionSignature(
  drawings: ChartDrawing[],
  fallback: Pick<DrawingScope, 'timeframe' | 'crossTimeframe'>,
  tombstones: DrawingTombstoneRevision[] = [],
) {
  const active = drawings.map((drawing) => {
    const location = drawingScope(drawing, fallback)
    return `a:${location.origin_timeframe}:${location.cross_timeframe ? '1' : '0'}:${drawing.id}:${drawing.revision ?? 0}`
  })
  const deleted = tombstones.map((tombstone) => `d:${drawingOperationKey(tombstone)}:${tombstone.revision}`)
  return [...active, ...deleted].sort().join('|')
}

export function mergeDrawingTombstoneRevisions(
  local: DrawingTombstone[],
  refreshed: DrawingTombstoneRevision[],
  latest: ChartDrawing[],
  fallback: Pick<DrawingScope, 'timeframe' | 'crossTimeframe'>,
) {
  const localByKey = new Map(local.map((tombstone) => [drawingOperationKey(tombstone), tombstone]))
  const latestByKey = new Map(latest.map((drawing) => {
    const location = drawingScope(drawing, fallback)
    return [drawingOperationKey({ drawing, ...location }), drawing] as const
  }))
  const merged: DrawingTombstone[] = []
  for (const revision of refreshed) {
    const key = drawingOperationKey(revision)
    const drawing = localByKey.get(key)?.drawing ?? latestByKey.get(key)
    if (!drawing) continue
    merged.push({
      drawing,
      revision: revision.revision,
      origin_timeframe: revision.origin_timeframe,
      cross_timeframe: revision.cross_timeframe,
    })
  }
  return merged
}

export function decideConflictRefresh(input: ConflictRefreshDecisionInput):
  | { kind: 'stale' | 'replace' | 'failed' }
  | { kind: 'retry'; retryKey: string } {
  if (input.capturedScopeToken !== input.currentScopeToken || input.capturedRequestGeneration !== input.currentRequestGeneration) return { kind: 'stale' }
  let pending = true
  try {
    pending = drawingOperations(input.refreshed, input.latest, input.fallback, input.tombstones).length > 0
  } catch { return { kind: 'failed' } }
  if (!pending) return { kind: 'replace' }
  const retryKey = `${input.currentScopeToken}:${input.currentEditGeneration}:${remoteRevisionSignature(input.refreshed, input.fallback, input.refreshedTombstones)}`
  return retryKey === input.previousRetryKey ? { kind: 'failed' } : { kind: 'retry', retryKey }
}

File: src/apps/api/chart_drawings.py
"""Account-scoped persistence for chart annotations.

The drawing payload is deliberately a small, closed schema.  It is not a
generic JSON document store: validating it here prevents one browser from
placing unbounded or ambiguous data in another user's chart scope.
"""

from __future__ import annotations

from datetime import date, datetime
import json
import math
import re
from typing import Any

from core.compat import UTC
from core.database import DatabaseManager, get_database
from src.apps.api.read_model import BrowserIdentity


MARKETS = {"US", "CN"}
SYMBOL_RE = re.compile(r"(?:[A-Z][A-Z0-9.-]{0,11}|\d{6})")
TIMEFRAME_RE = re.compile(r"[\w\-]{1,16}", re.UNICODE)
UUID_RE = re.compile(r"[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}", re.I)
DATE_RE = re.compile(r"\d{4}-\d{2}-\d{2}")
ACTIVE_DRAWING_LIMIT = 200
USER_TOMBSTONE_LIMIT = 2_000
SYMBOL_TOMBSTONE_LIMIT = 500

TOOL_POINTS = {
    "segment": 2, "horizontal": 1, "horizontal-segment": 2, "vertical": 1,
    "ray": 2, "straight": 2, "parallel": 3, "channel": 3, "periodic": 2,
    "info-line": 2, "smooth-top": 3, "cross": 1, "rectangle": 2,
    "triangle": 3, "parallelogram": 3, "circle": 2, "ellipse": 2, "path": 5,
    "wave3": 4, "wave5": 6, "wave8": 9, "head-shoulders": 5,
    "triangle-pattern": 5, "mw": 5, "abcd": 4, "xabcd": 5,
    "three-drive": 7, "sine": 6, "fib-retracement": 2, "fib-time": 2,
    "fib-extension": 3, "speed-resistance": 2, "gann-box": 2,
    "gann-angle": 2, "grid-line": 2, "pitchfork": 3, "schiff": 3,
    "modified-schiff": 3, "inside-pitchfork": 3, "fan": 2, "time-ruler": 2,
    "space-ruler": 2, "time-space-ruler": 2, "long-position": 2,
    "short-position": 2, "price-label": 1, "arrow": 2, "up-arrow": 2,
    "down-arrow": 2,
}


class ChartDrawingError(ValueError):
    pass


class ChartDrawingConflict(ChartDrawingError):
    pass


def _now() -> str:
    return datetime.now(UTC).isoformat(timespec="seconds")


def _scope(payload: dict[str, Any]) -> tuple[str, str, str, int]:
    if set(payload) != {"market", "symbol", "timeframe", "cross_timeframe"}:
        raise ChartDrawingError("画线查询字段无效。")
    market = payload["market"]
    symbol = payload["symbol"]
    timeframe = payload["timeframe"]
    cross = payload["cross_timeframe"]
    if not isinstance(market, str) or market not in MARKETS:
        raise ChartDrawingError("画线市场必须是 US 或 CN。")
    if not isinstance(symbol, str) or not SYMBOL_RE.fullmatch(symbol.upper()):
        raise ChartDrawingError("画线股票代码无效。")
    if not isinstance(timeframe, str) or not TIMEFRAME_RE.fullmatch(timeframe):
        raise ChartDrawingError("画线周期无效。")
    if not isinstance(cross, bool):
        raise ChartDrawingError("跨周期标志必须为布尔值。")
    return market, symbol.upper(), timeframe, int(cross)


def query_scope(*, market: Any, symbol: Any, timeframe: Any, cross_timeframe: Any) -> tuple[str, str, str, int]:
    return _scope({"market": market, "symbol": symbol, "timeframe": timeframe, "cross_timeframe": cross_timeframe})


def _drawing(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict) or set(value) != {"id", "tool", "points"}:
        raise ChartDrawingError("画线对象包含未知字段。")
    drawing_id, tool, points = value["id"], value["tool"], value["points"]
    if not isinstance(drawing_id, str) or not UUID_RE.fullmatch(drawing_id):
        raise ChartDrawingError("画线编号无效。")
    if not isinstance(tool, str) or tool not in TOOL_POINTS:
        raise ChartDrawingError("画线工具无效。")
    if not isinstance(points, list) or len(points) != TOOL_POINTS[tool]:
        raise ChartDrawingError("画线点数量无效。")
    normalized: list[dict[str, Any]] = []
    for point in points:
        if not isinstance(point, dict) or set(point) != {"time", "price"}:
            raise ChartDrawingError("画线点包含未知字段。")
        time, price = point["time"], point["price"]
        business_day = isinstance(time, dict) and set(time) == {"year", "month", "day"}
        if business_day:
            try:
                if any(isinstance(time[key], bool) or not isinstance(time[key], int) for key in time):
                    raise ValueError
                date(time["year"], time["month"], time["day"])
            except (TypeError, ValueError):
                business_day = False
        valid_date_string = False
        if isinstance(time, str) and DATE_RE.fullmatch(time):
            try:
                date.fromisoformat(time)
                valid_date_string = True
            except ValueError:
                pass
        if isinstance(time, bool) or not (
            isinstance(time, int) and 0 < time < 4_102_444_800
            or valid_date_string
            or business_day
        ):
            raise ChartDrawingError("画线时间无效。")
        if isinstance(price, bool) or not isinstance(price, (int, float)) or not math.isfinite(float(price)) or abs(float(price)) > 1_000_000_000:
            raise ChartDrawingError("画线价格无效。")
        normalized.append({"time": time, "price": float(price)})
    return {"id": drawing_id.lower(), "tool": tool, "points": normalized}


def _operation(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict) or not isinstance(value.get("op"), str):
        raise ChartDrawingError("画线操作无效。")
    op = value["op"]
    if op == "upsert":
        if set(value) != {"op", "origin_timeframe", "cross_timeframe", "revision", "drawing"}:
            raise ChartDrawingError("保存画线操作字段无效。")
        drawing = _drawing(value["drawing"])
        origin_timeframe, cross_timeframe = _operation_scope(value)
        revision = value["revision"]
        if revision is not None and (isinstance(revision, bool) or not isinstance(revision, int) or revision < 1):
            raise ChartDrawingError("画线版本无效。")
        return {"op": op, "origin_timeframe": origin_timeframe, "cross_timeframe": cross_timeframe, "revision": revision, "drawing": drawing, "drawing_id": drawing["id"]}
    if op in {"delete", "restore"}:
        if set(value) != {"op", "origin_timeframe", "cross_timeframe", "revision", "drawing_id"}:
            raise ChartDrawingError("画线操作字段无效。")
        origin_timeframe, cross_timeframe = _operation_scope(value)
        drawing_id, revision = value["drawing_id"], value["revision"]
        if not isinstance(drawing_id, str) or not UUID_RE.fullmatch(drawing_id):
            raise ChartDrawingError("画线编号无效。")
        if isinstance(revision, bool) or not isinstance(revision, int) or revision < 1:
            raise ChartDrawingError("画线版本无效。")
        return {"op": op, "origin_timeframe": origin_timeframe, "cross_timeframe": cross_timeframe, "revision": revision, "drawing_id": drawing_id.lower()}
    raise ChartDrawingError("不支持的画线操作。")


def _operation_scope(value: dict[str, Any]) -> tuple[str, int]:
    timeframe, cross = value.get("origin_timeframe"), value.get("cross_timeframe")
    if not isinstance(timeframe, str) or not TIMEFRAME_RE.fullmatch(timeframe):
        raise ChartDrawingError("画线来源周期无效。")
    if not isinstance(cross, bool):
        raise ChartDrawingError("跨周期标志必须为布尔值。")
    return timeframe, int(cross)


class ChartDrawingService:
    def __init__(self, database: DatabaseManager | None = None):
        self.db = database or get_database()

    def list(self, identity: BrowserIdentity, *, market: Any, symbol: Any, timeframe: Any, cross_timeframe: Any) -> dict[str, Any]:
        market, symbol, timeframe, cross = query_scope(market=market, symbol=symbol, timeframe=timeframe, cross_timeframe=cross_timeframe)
        if cross:
            rows = self.db.fetch_all(
                """SELECT origin_timeframe,cross_timeframe,drawing_json,revision FROM user_chart_drawings
                   WHERE user_id=? AND market=? AND symbol=? AND deleted_at IS NULL
                   AND ((cross_timeframe=0 AND origin_timeframe=?) OR cross_timeframe=1)
                   ORDER BY cross_timeframe,origin_timeframe,created_at LIMIT 201""",
                (identity.id, market, symbol, timeframe),
            )
            deleted_rows = self.db.fetch_all(
                """SELECT origin_timeframe,cross_timeframe,drawing_id,revision FROM user_chart_drawings
                   WHERE user_id=? AND market=? AND symbol=? AND deleted_at IS NOT NULL
                   AND ((cross_timeframe=0 AND origin_timeframe=?) OR cross_timeframe=1)
                   ORDER BY updated_at DESC LIMIT ?""",
                (identity.id, market, symbol, timeframe, SYMBOL_TOMBSTONE_LIMIT + 1),
            )
        else:
            rows = self.db.fetch_all(
                """SELECT origin_timeframe,cross_timeframe,drawing_json,revision FROM user_chart_drawings
                   WHERE user_id=? AND market=? AND symbol=? AND origin_timeframe=?
                   AND cross_timeframe=0 AND deleted_at IS NULL ORDER BY created_at LIMIT 201""",
                (identity.id, market, symbol, timeframe),
            )
            deleted_rows = self.db.fetch_all(
                """SELECT origin_timeframe,cross_timeframe,drawing_id,revision FROM user_chart_drawings
                   WHERE user_id=? AND market=? AND symbol=? AND origin_timeframe=?
                   AND cross_timeframe=0 AND deleted_at IS NOT NULL
                   ORDER BY updated_at DESC LIMIT ?""",
                (identity.id, market, symbol, timeframe, SYMBOL_TOMBSTONE_LIMIT + 1),
            )
        items = []
        for row in rows[:ACTIVE_DRAWING_LIMIT]:
            drawing = json.loads(row["drawing_json"])
            items.append({**drawing, "origin_timeframe": row["origin_timeframe"], "cross_timeframe": bool(row["cross_timeframe"]), "revision": row["revision"]})
        tombstones = [
            {
                "drawing_id": row["drawing_id"],
                "origin_timeframe": row["origin_timeframe"],
                "cross_timeframe": bool(row["cross_timeframe"]),
                "revision": row["revision"],
            }
            for row in deleted_rows[:SYMBOL_TOMBSTONE_LIMIT]
        ]
        return {
            "items": items,
            "truncated": len(rows) > ACTIVE_DRAWING_LIMIT,
            "tombstones": tombstones,
            "tombstones_truncated": len(deleted_rows) > SYMBOL_TOMBSTONE_LIMIT,
        }

    def batch(self, identity: BrowserIdentity, payload: dict[str, Any]) -> dict[str, Any]:
        if set(payload) != {"market", "symbol", "operations"}:
            raise ChartDrawingError("画线批量请求字段无效。")
        market, symbol = payload["market"], payload["symbol"]
        if not isinstance(market, str) or market not in MARKETS:
            raise ChartDrawingError("画线市场必须是 US 或 CN。")
        if not isinstance(symbol, str) or not SYMBOL_RE.fullmatch(symbol.upper()):
            raise ChartDrawingError("画线股票代码无效。")
        symbol = symbol.upper()
        operations = payload["operations"]
        if not isinstance(operations, list) or not operations or len(operations) > 100:
            raise ChartDrawingError("每批画线操作必须介于 1 与 100 条。")
        normalized = [_operation(item) for item in operations]
        if len({(item["origin_timeframe"], item["cross_timeframe"], item["drawing_id"]) for item in normalized}) != len(normalized):
            raise ChartDrawingError("同一批不能重复操作同一画线。")
        now = _now()
        results: list[dict[str, Any]] = []
        with self.db.transaction() as connection:
            connection.execute("BEGIN IMMEDIATE")
            for operation in normalized:
                row = connection.execute(
                    """SELECT drawing_json,revision,deleted_at FROM user_chart_drawings
                       WHERE user_id=? AND market=? AND symbol=? AND origin_timeframe=? AND cross_timeframe=? AND drawing_id=?""",
                    (identity.id, market, symbol, operation["origin_timeframe"], operation["cross_timeframe"], operation["drawing_id"]),
                ).fetchone()
                if operation["op"] == "upsert":
                    encoded = json.dumps(operation["drawing"], ensure_ascii=False, separators=(",", ":"))
                    if row is None:
                        if operation["revision"] is not None:
                            raise ChartDrawingConflict("画线版本冲突，请重新读取。")
                        self._require_active_capacity(connection, identity.id, market, symbol)
                        connection.execute(
                            """INSERT INTO user_chart_drawings(user_id,market,symbol,origin_timeframe,cross_timeframe,drawing_id,drawing_json,revision,deleted_at,created_at,updated_at)
                               VALUES (?,?,?,?,?,?,?,?,NULL,?,?)""",
                            (identity.id, market, symbol, operation["origin_timeframe"], operation["cross_timeframe"], operation["drawing_id"], encoded, 1, now, now),
                        )
                        revision, deleted_at = 1, None
                    elif row["deleted_at"] is None and row["drawing_json"] == encoded and operation["revision"] in {None, row["revision"], row["revision"] - 1}:
                        revision, deleted_at = row["revision"], None
                    else:
                        if operation["revision"] != row["revision"]:
                            raise ChartDrawingConflict("画线版本冲突，请重新读取。")
                        if row["deleted_at"] is not None:
                            self._require_active_capacity(connection, identity.id, market, symbol)
                        revision, deleted_at = row["revision"] + 1, None
                        connection.execute(
                            """UPDATE user_chart_drawings SET drawing_json=?,revision=?,deleted_at=NULL,updated_at=?
                               WHERE user_id=? AND market=? AND symbol=? AND origin_timeframe=? AND cross_timeframe=? AND drawing_id=?""",
                            (encoded, revision, now, identity.id, market, symbol, operation["origin_timeframe"], operation["cross_timeframe"], operation["drawing_id"]),
                        )
                else:
                    if row is None:
                        raise ChartDrawingConflict("画线不存在或版本冲突，请重新读取。")
                    desired_deleted = operation["op"] == "delete"
                    already_desired = (row["deleted_at"] is not None) == desired_deleted
                    if already_desired and operation["revision"] in {row["revision"], row["revision"] - 1}:
                        revision, deleted_at = row["revision"], row["deleted_at"]
                    else:
                        if operation["revision"] != row["revision"]:
                            raise ChartDrawingConflict("画线版本冲突，请重新读取。")
                        if not desired_deleted and row["deleted_at"] is not None:
                            self._require_active_capacity(connection, identity.id, market, symbol)
                        revision = row["revision"] + 1
                        deleted_at = now if desired_deleted else None
                        connection.execute(
                            """UPDATE user_chart_drawings SET revision=?,deleted_at=?,updated_at=?
                               WHERE user_id=? AND market=? AND symbol=? AND origin_timeframe=? AND cross_timeframe=? AND drawing_id=?""",
                            (revision, deleted_at, now, identity.id, market, symbol, operation["origin_timeframe"], operation["cross_timeframe"], operation["drawing_id"]),
                        )
                results.append({"drawing_id": operation["drawing_id"], "origin_timeframe": operation["origin_timeframe"], "cross_timeframe": bool(operation["cross_timeframe"]), "revision": revision, "deleted": deleted_at is not None})
            connection.execute(
                """DELETE FROM user_chart_drawings WHERE rowid IN (
                       SELECT rowid FROM user_chart_drawings
                       WHERE user_id=? AND market=? AND symbol=? AND deleted_at IS NOT NULL
                       ORDER BY updated_at DESC LIMIT -1 OFFSET ?
                   )""",
                (identity.id, market, symbol, SYMBOL_TOMBSTONE_LIMIT),
            )
            connection.execute(
                """DELETE FROM user_chart_drawings WHERE rowid IN (
                       SELECT rowid FROM user_chart_drawings WHERE user_id=? AND deleted_at IS NOT NULL
                       ORDER BY updated_at DESC LIMIT -1 OFFSET ?
                   )""",
                (identity.id, USER_TOMBSTONE_LIMIT),
            )
        return {"items": results}

    @staticmethod
    def _require_active_capacity(connection: Any, user_id: int, market: str, symbol: str) -> None:
        count = connection.execute(
            """SELECT COUNT(*) FROM user_chart_drawings
               WHERE user_id=? AND market=? AND symbol=? AND deleted_at IS NULL""",
            (user_id, market, symbol),
        ).fetchone()[0]
        if count >= ACTIVE_DRAWING_LIMIT:
            raise ChartDrawingError(f"每个股票最多保存 {ACTIVE_DRAWING_LIMIT} 条有效画线。")
